include/glow/Runtime/Executor/NetworkExecutionState.h

```diff
@@ -0,0 +1,156 @@
+  /// called at the same time as getRawResultPlaceholderBindingsPtr() or
+  /// insertIntoResultCtx().
+  std::unique_ptr<ExecutionContext> getUniqueResultContextPtr();
+
+  /// \returns a raw pointer to the result bindings. This should be not called
+  /// at the same time as getUniqueResultPlaceholderBindingsPtr().
+  ExecutionContext *getRawResultContextPtr() const;
+
+  /// \returns the callback for this execution.
+  ResultCBTy getCallback() { return cb_; }
+
+  /// \returns the OneErrOnly Error container for the execution.
+  OneErrOnly &getErrorContainer() { return errContainer_; }
+
+  /// \returns the run ID for the execution.
+  RunIdentifierTy getRunId() const { return runId_; }
+
+  /// \returns the Executor used for this run.
+  // ThreadExecutor *getExecutor() { return executor_; }
+
+  /// Whether or not this node has been initialized.
+  bool initialized_{false};
+
+private:
+  /// The run identifier for this execution of a DAG.
+  RunIdentifierTy runId_;
+  /// The callback that should be called when execution is done.
+  ResultCBTy cb_;
+
+  /// Map of all buffers allocated for intermediate contexts.
+  std::unordered_map<Placeholder *, void *> buffers_;
+
+  /// Map from buffer to device that allocated it, used at destruction to
+  /// free buffers.
+  std::unordered_map<void *, DeviceManager *> deviceAllocations_;
+
+  /// Map of intermediate placeholder bindings that need to be pointed at
+  /// resultCtx tensors.
+  std::unordered_map<Placeholder *, std::vector<PlaceholderBindings *>>
+      externalIntermediates_;
+
+  /// The ExecutionContext object containing the results of the execution
+  /// (i.e. the outputs of the DAGNodes that have no children).
+  std::unique_ptr<ExecutionContext> resultCtx_;
+  /// Counters for how many of each nodes parents are done. These are needed
+  /// in order to determine when a node is ready to be executed.
+  std::unordered_map<const DAGNode *, std::atomic<unsigned>> nodeParentsDone_;
+  /// Input contexts for all of the nodes. These are gradually
+  /// populated as a node's parents finish.
+  std::unordered_map<const DAGNode *, std::unique_ptr<ExecutionContext>>
+      intermediateContexts_;
+  /// This is populated with the roots when a run starts, and does not become
+  /// empty until execution finishes.
+  std::atomic<unsigned> inflightNodes_;
+  /// Value that is used to track if an Error was received.
+  OneErrOnly errContainer_;
+  /// Module for the network. This contains the PHs used by the functions in
+  /// this network.
+  Module *module_{nullptr};
+  /// Root node of the DAG for this run.
+  const DAGNode *root_;
+
+  /// Thread Executor used for this run.
+  // ThreadExecutor *executor_;
+};
+
+class NetworkExecutionStatePool {
+public:
+  NetworkExecutionState *getNextNetworkExecutionState() {
+    unsigned current = currentState_;
+    currentState_ = (currentState_ + 1) % states_.size();
+    return states_[current].get();
```

Comment:
This is not thread safe. Two threads can read the same currentState_. Even with currentState_.fetch_and_add() it is still not safe. Here you can do what we did for the NNPI EnvPool, using a stack guarded by mutex. Or since we have access to boost now, we can use boost::lock_free::stack<NetExecutionState>.

Reply:
I'm hoping to avoid a mutex since this is in the inference path, If I use fetch_add(), and move the modulus to line 142, It should be thread safe, states_.size() should not change after initialization. I can refactor create pool to make that more explicit.

Reply:
Modulus will wrap around.

Reply:
That's fine, since the size of the pool == maxActiveRequests. there won't be overlap.

Reply:
It's not safe. As I mentioned, having pool size equal to maxActiveRequests may not be a good idea. And relying on this implicitly may potentially lead to hard to debug issues. Mutex is actually not that bad and we have access to lockfree stack also. Such scheme has been proven to be efficient in large scale services.